ws/src/main/scala/de/thm/ii/fbs/services/evaluation/EvaluationResultService.scala

```diff
@@ -0,0 +1,71 @@
+package de.thm.ii.fbs.services.evaluation
+
+import de.thm.ii.fbs.model.{CourseEvaluationResult, CourseResult, EvaluationContainer,
+  EvaluationContainerResult, EvaluationContainerWithTaskResults, Task, TaskResult}
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+import scala.math.BigDecimal.RoundingMode
+
+/**
+  * Evaluate Container results
+  */
+@Component
+class EvaluationResultService {
+  @Autowired
+  private val formulaService: FormulaService = null
+
+  /**
+    * Evaluate Course results
+    * @param container Evaluation Container
+    * @param results the Course Results
+    * @return Evaluation Course results for a User
+    */
+  def evaluate(container: List[EvaluationContainer], results: List[CourseResult]): List[CourseEvaluationResult] = {
+    results.map(r => evaluateOne(container, r))
+  }
+
+  private def evaluateOne(container: List[EvaluationContainer], result: CourseResult): CourseEvaluationResult = {
+    val taskResults = result.results.foldLeft(Map[Int, TaskResult]()) { (m, s) => m + (s.task.id -> s) }
+    var passed = true
+    var bonusPoints = 0
+
+    val evaluationResult = container.map(c => {
+      val (passedTasks, tasksResults) = getTaskResults(c.tasks, taskResults)
+      val cPassed = wasPassed(passedTasks, c.toPass)
+      val cBonusPoints = calculateBonusPoints(c.bonusFormula, passedTasks, c.toPass)
+
+      bonusPoints += cBonusPoints
+      if (!cPassed) passed = false
+
+      EvaluationContainerResult(cPassed, cBonusPoints, passedTasks,
+        EvaluationContainerWithTaskResults(c.id, tasksResults, c.toPass, c.bonusFormula, c.hidePoints))
+    })
+
+    CourseEvaluationResult(result.user, passed, bonusPoints, evaluationResult)
+  }
+
+  private def getTaskResults(tasks: List[Task], taskResults: Map[Int, TaskResult]) = {
+    var passedTasks = 0
+    val tasksResults = tasks.map(t => {
+      val tRes = taskResults(t.id)
+      if (tRes.passed) passedTasks += 1
+      tRes
+    })
+
+    (passedTasks, tasksResults)
+  }
+
+  private def wasPassed(passedTasks: Int, toPass: Int) = passedTasks >= toPass
+
+  private def calculateBonusPoints(bonusFormula: String, passedTasks: Int, toPass: Int) = {
+    val bonusPointsVariables = Map("x" -> passedTasks.toString, "y" -> toPass.toString)
+
+    try {
+      if (bonusFormula != null) formulaService.evaluate(bonusFormula, bonusPointsVariables).setScale(0, RoundingMode.HALF_UP).toInt else 0
+    } catch {
+      // If the bonus formula is invalid, the entire calculation should not fail
+      case _: Exception => 0
+    }
```

Comment:
- Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?  
- If the bonus formula calculation results in a negative number, you have negative bonus points, should we change this to a `0`?

Reply:
> * Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?

if easily possible i think we should change the datatype to float to allow half ponus points.

> * If the bonus formula calculation results in a negative number, you have negative bonus points, should we change this to a `0`?

I think that negative bonus points do not make sense, so we should set this to zero if negative.

Reply:
> if easily possible i think we should change the datatype to float to allow half ponus points.

The implementation in the backend should not be very complicated. @kimiannalena  how complex would this change be in the frontend?

> I think that negative bonus points do not make sense, so we should set this to zero if negative.

:+1:

Reply:
> > Currently, the result of the bonus formula is rounded mathematically, as bonus points have the data type 'Int'. Should we leave it like this?

> if easily possible i think we should change the datatype to float to allow half ponus points.

I think that half points are not relevant for lecturers. At least I have never encountered any.
I would suggest to let the lecturer choose the method of rounding (i.e. mathematically, up or down.
As an example, in database systems we require students to solve 4 tasks to acquire one bonus point. If a student only solves 3, they do not get a point.

Reply:
> I think that half points are not relevant for lecturers. At least I have never encountered any.
> I suggest to let the lecturer choose the method of rounding (i.e. mathematically, up or down.
> As an example, in database systems we require students to solve 4 tasks to acquire one bonus point. If a student only solves 3, they do not get a point.

This is possible using functions in the formula (e.g. `FLOOR` to round the value to the nearest integer. A list of all functions can be found [here](https://github.com/uklimaschewski/EvalEx#supported-functions)). So if you want to use a rounding mode other than the default (matematic) you can, if you don't the number will automatically be rounded matematically to get an integer.

> I think that negative bonus points do not make sense, so we should set this to zero if negative.

@Zitrone44 Negative numbers are now converted to `0`.